Add audio track selection to editor setup screen

peppy opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/10233

Drag & drop will come later.

peppy committed:
Add basic testing of new beatmaps persistence

peppy committed:
Add audio selection to song setup screen

peppy committed:
Add ability to actually import a new audio file to the beatmap / database

peppy committed:
Remove the previous audio file before adding a new one

peppy committed:
Add basic track reloading support while inside the editor

peppy committed:
Use bindable for track to fix rate adjustments not applying correctly

peppy committed:
Fix test failures due to editor dependency

peppy committed:
Fix summary timeline not updating to new track length correctly

peppy committed:
Add test coverage of audio track changing

peppy committed:
Fix test runs under headless

peppy committed:
Remove weird using

bdach commented on osu.Game.Tests/Visual/Editing/TestSceneEditorBeatmapCreation.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        protected override Ruleset CreateEditorRuleset() => new OsuRuleset();
+
+        protected override bool EditorComponentsReady => Editor.ChildrenOfType<SetupScreen>().FirstOrDefault()?.IsLoaded == true;
```
as is tradition
```suggestion
        protected override bool EditorComponentsReady => Editor.ChildrenOfType<SetupScreen>().SingleOrDefault()?.IsLoaded == true;
```

bdach commented on osu.Game.Tests/Visual/Editing/TestSceneEditorBeatmapCreation.cs:
```diff
@@ -0,0 +1,66 @@
+        [Test]
+        public void TestCreateNewBeatmap()
+        {
+            AddStep("add random hitobject", () => EditorBeatmap.Metadata.Title = Guid.NewGuid().ToString());
```
step name and action are two different things :wink: not sure which way you meant to go here

peppy replied:
My line of thought was "this test is failing from saving multiple non-unique beatmaps" so I was going to add a hitobject. Then realised changing the metadata was much simpler. Oops.

bdach commented on osu.Game/Overlays/MusicController.cs:
```diff
@@ -81,6 +81,11 @@ private void load()
+        /// <summary>
+        /// Forcefully reload the current <see cref="WorkingBeatmap"/>'s track from disk.
+        /// </summary>
+        public void ForceReloadCurrentBeatmap() => changeTrack();
```
Would probably rename to `ForceReloadCurrentTrack`. Reloading beatmap in music controller sounds a little misleading.

bdach commented on osu.Game/Screens/Edit/Setup/SetupScreen.cs:
```diff
@@ -120,4 +202,60 @@ private void onCommit(TextBox sender, bool newText)
+
+        public void DisplayFileChooser()
+        {
+            Target.Child = new FileSelector("/Users/Dean/.osu/Songs", new[] { ".mp3", ".ogg" })
```
not the best default for anyone else :stuck_out_tongue: Should probably be home dir if starting from scratch/editing existing, or latest song location if one has been chosen before in current editor session?

bdach commented on osu.Game/Screens/Edit/Setup/SetupScreen.cs:
```diff
@@ -104,10 +140,56 @@ private void load(OsuColour colours)
+
+            music.ForceReloadCurrentBeatmap();
+
+            editor.UpdateClockSource();
```
`editor` is permitted to be null
```suggestion
            editor?.UpdateClockSource();
```

peppy replied:
Yeah, it is, but I didn't feel the need to fix this since it's likely never going to be called with a null editor. I guess there's no harm adding the check, though.

peppy committed:
Merge branch 'master' into editor-load-audio

peppy committed:
Use SingleOrDefault where feasible

peppy committed:
Fix misleading step name (and add comment as to its purpose)

peppy committed:
Rename reload method to not mention beatmap unnecessarily

peppy committed:
Add null check because we can

peppy committed:
Remove local specification for file selector search path

peppy committed:
Merge branch 'file-selector' into editor-load-audio

peppy committed:
Add "bindables" to dictionary

peppy commented:
I'm in the process of moving the file operations out into `WorkingBeatmap` / `BeatmapManager` for what it's worth.

The actually process won't change but I may push to this PR (or can just leave for a follow-up PR).

peppy committed:
Expand available file operations in ArchiveModelManager

peppy committed:
Remove local file handling from SetupScreen

peppy commented:
Ended up making the changes on this branch since they weren't as large as I expected, should be good to go.

smoogipoo committed:
Merge branch 'master' into editor-load-audio

peppy committed:
Merge branch 'master' into editor-load-audio

peppy committed:
Fix test failures

The issue was the ArchiveModelManager change; the test local change is
just there because it makes more sense to run for every test in that
scene.

peppy committed:
Merge branch 'master' into editor-load-audio
